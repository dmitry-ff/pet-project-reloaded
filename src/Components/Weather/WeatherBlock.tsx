import _, { isEmpty } from 'lodash';
import React from 'react';
import { WEEK_S } from "../../Data/Consts/WEEK_S";
import { WeatherData, WeatherToday, FirstColumn, SecondColumn, FourthColumn, WeatherWeek, WeatherCard } from '../../Style/Weather'
import { TLoading } from "../../Data/Types/TLoading";
import { EmotionJSX } from '@emotion/react/types/jsx-namespace';
type THead = {
  nameHead: string;
}
type TData = {
  data: {
    date: string,
    day: {
      maxtemp_c: number,
      mintemp_c: number,
      avgtemp_c: number,
      maxwind_kph: number,
      daily_chance_of_rain: number,
      daily_chance_of_snow: number,
      condition: {
        text: string,
        icon: string
      }
    }
  }[]
}

const WeatherBlock: React.FC<TLoading & THead & TData> = ({ loading, nameHead, data }): EmotionJSX.Element | null => {
  console.log(loading, 'LOADING')
  console.log(data, 'DATA');

  if (!loading && isEmpty(data)) return null

  return (
    <WeatherData>
      <span >
        Weather in <b>{nameHead}</b>
      </span>
      <WeatherToday>
        <FirstColumn>
          <span>{_.head(data)!.day.avgtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.avgtemp_c) : _.floor(_.head(data)!.day.avgtemp_c)}°</span>
          <img src={_.head(data)!.day.condition.icon} alt={_.head(data)!.day.condition.text} />
        </FirstColumn>
        <SecondColumn>
          <span>{_.head(data)!.day.condition.text}</span>
          <div>
            <span>max <b>{_.head(data)!.day.maxtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.maxtemp_c) : _.floor(_.head(data)!.day.maxtemp_c)}</b></span>
            <span>min <b>{_.head(data)!.day.mintemp_c > 0 ? '+' + _.floor(_.head(data)!.day.mintemp_c) : _.floor(_.head(data)!.day.mintemp_c)}</b></span>
          </div>
        </SecondColumn>
        <div>
          <span>Wind speed  <b>{(_.head(data)!.day.maxwind_kph / 3.6).toFixed(1)} m/s</b></span>
        </div>
        <FourthColumn>
          <span>{_.head(data)!.day.daily_chance_of_rain} %</span>
          <span>{_.head(data)!.day.daily_chance_of_snow} %</span>
        </FourthColumn>
      </WeatherToday>

      <WeatherWeek>
        {!loading && <>
          {
            _.map(data, (item, index) =>
              <div key={index}>
                <div>
                  {index === 0 ?
                    'Today'
                    : `${WEEK_S[new Date(item.date).getDay()]}, ${new Date(item.date).getDate()} `}
                </div>
                <WeatherCard>
                  <div>
                    <span>{item.day.maxtemp_c > 0 ? '+' + _.floor(item.day.maxtemp_c) : _.floor(item.day.maxtemp_c)}°</span>
                    <span>{item.day.mintemp_c > 0 ? '+' + _.floor(item.day.mintemp_c) : _.floor(item.day.mintemp_c)}°</span>
                  </div>
                  <img src={item.day.condition.icon} alt={item.day.condition.text} />
                </WeatherCard>
              </div>)
          }
        </>
        }

      </WeatherWeek>
    </WeatherData>
  )

}

export default WeatherBlock;
// export { }