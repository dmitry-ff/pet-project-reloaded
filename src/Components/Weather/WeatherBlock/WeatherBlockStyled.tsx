import { css } from '@emotion/react';
import styled from '@emotion/styled';
import rain from '../../../Images/7795646_weather_cloud_rain_icon.svg';
import snow from '../../../Images/8664916_snowflake_icon.svg';
export const WeatherData = styled.div`
    margin-bottom: 35px;
    border-radius: 10px;
    padding:20px;
    box-shadow: 0 0 10px rgba(219, 219, 219, 0.596);
    display: flex;
    margin:auto;
    flex-direction: column;
    align-content:center;
    height: 300px;
    width:max-content;
    min-width:600px;
    justify-content: space-between;
    >span{
        font-size:1.5em;
    }
  
`

export const WeatherToday = styled.div`
    display: flex;
    align-items:flex-end ;
    justify-content: space-between;
    div {
        margin-right:20px;
        display:flex;
    }
`
export const FirstColumn = styled.div`
    display: flex;
    align-items: flex-end;
    & > span{
        font-size: 3em;
    }

`

export const SecondColumn = styled.div`
    display: flex;
    flex-direction: column;
    div{
        display: flex;
        justify-content: space-between;
        width: 150px;
    }
`

const baseStyle = css`
    position: absolute;
    content:'';
    right:-20px;
    
    display: inline-block;
    width: 17px;
    height: 17px;
    background-repeat: no-repeat;
    padding: 0;
`;

export const FourthColumn = styled.div`
    display: flex;
    flex-direction: column;
    align-items:flex-end;
    justify-content:space-between;

    & :first-of-type {
        position: relative;
        &:after{
            vertical-align: top;
            background-image:url(${rain});
            ${baseStyle};
        }
    }
    & :last-of-type {
        position: relative;
        &:after{
            vertical-align: center;
            background-image: url(${snow});
            ${baseStyle};
        }
    }
`


export const WeatherWeek = styled.div`
    display: flex;
    justify-content: space-evenly;
    > div{
        width: 100px;
        border-radius: 15px;
        padding: 15px 20px;
        box-shadow: 5px 5px 10px rgba(211, 211, 211, 0.712);
        background-color: rgba(225, 241, 245, 0.705);
    }
`

export const WeatherCard = styled.div`
        display: flex;
        > div{
            justify-content: space-around;
            display: flex;
            flex-direction: column;
            & :first-of-type{
               font-size:1.5em;
            }
            & :last-of-type{
                font-size:0.8em;   
                color: rgb(102, 102, 102);
            }
        }

`