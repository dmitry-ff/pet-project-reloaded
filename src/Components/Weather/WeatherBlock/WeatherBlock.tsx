import _, { isEmpty } from 'lodash';
import { WEEK_S } from "../data/WEEK_S";
import * as Styled from './WeatherBlock.styled'
import { EmotionJSX } from '@emotion/react/types/jsx-namespace';
import { TData } from '../types/TData'



const WeatherBlock = (props: { loading: boolean, nameHead: string, data: TData[] }): EmotionJSX.Element | null => {
  const { loading, nameHead, data } = props;
  if (!loading && isEmpty(data)) return null;

  return (
    <Styled.WeatherData>
      <span >
        Weather in <b>{nameHead}</b>
      </span>
      <Styled.WeatherToday>
        <Styled.FirstColumn>
          <span>{_.head(data)!.day.avgtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.avgtemp_c) : _.floor(_.head(data)!.day.avgtemp_c)}°</span>
          <img src={_.head(data)!.day.condition.icon} alt={_.head(data)!.day.condition.text} />
        </Styled.FirstColumn>
        <Styled.SecondColumn>
          <span>{_.head(data)!.day.condition.text}</span>
          <div>
            <span>max <b>{_.head(data)!.day.maxtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.maxtemp_c) : _.floor(_.head(data)!.day.maxtemp_c)}</b></span>
            <span>min <b>{_.head(data)!.day.mintemp_c > 0 ? '+' + _.floor(_.head(data)!.day.mintemp_c) : _.floor(_.head(data)!.day.mintemp_c)}</b></span>
          </div>
        </Styled.SecondColumn>
        <div>
          <span>Wind speed  <b>{(_.head(data)!.day.maxwind_kph / 3.6).toFixed(1)} m/s</b></span>
        </div>
        <Styled.FourthColumn>
          <span>{_.head(data)!.day.daily_chance_of_rain} %</span>
          <span>{_.head(data)!.day.daily_chance_of_snow} %</span>
        </Styled.FourthColumn>
      </Styled.WeatherToday>

      <Styled.WeatherWeek>
        {!loading && <>
          {
            _.map(data, (item, index) =>
              <div key={index}>
                <div>
                  {index === 0 ?
                    'Today'
                    : `${WEEK_S[new Date(item.date).getDay()]}, ${new Date(item.date).getDate()} `}
                </div>
                <Styled.WeatherCard>
                  <div>
                    <span>{item.day.maxtemp_c > 0 ? '+' + _.floor(item.day.maxtemp_c) : _.floor(item.day.maxtemp_c)}°</span>
                    <span>{item.day.mintemp_c > 0 ? '+' + _.floor(item.day.mintemp_c) : _.floor(item.day.mintemp_c)}°</span>
                  </div>
                  <img src={item.day.condition.icon} alt={item.day.condition.text} />
                </Styled.WeatherCard>
              </div>)
          }
        </>
        }

      </Styled.WeatherWeek>
    </Styled.WeatherData>
  )

}

export default WeatherBlock;
// export { }