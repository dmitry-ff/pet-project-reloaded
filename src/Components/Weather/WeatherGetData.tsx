import React from 'react';
import axios from 'axios';
import TemperatureChart from './TemperatureChart';
import { CircularProgress, Box } from '@mui/material';
import { TResponse } from "../../Data/Types/TResponse";
import { TData } from "../../Data/Types/TData";
import { TContext } from "../../Data/Types/TContext";
import { Context } from '../../Context'
import WeatherBlock from './WeatherBlock';
import { EmotionJSX } from '@emotion/react/types/jsx-namespace';
import styled from '@emotion/styled';

type TProps = {
  apiURL: string;
  townName: string;
}
const ErrorBlock = styled.div`
  padding:10px 45px;
  background-color:rgba(240, 15, 71,0.1);
  border:1px solid rgb(246, 66, 95);
  border-radius:10px;
  margin-bottom:15px;

`

function WeatherGetData({ apiURL, townName }: TProps): EmotionJSX.Element {

  const [data, setData] = React.useState<TData[]>([]);
  const [isLoading, setLoading] = React.useState<boolean>(true);
  const [isError, setIsError] = React.useState<boolean>(false)

  React.useEffect(() => {
    axios.get<TResponse>(apiURL)
      .then(res => {
        setLoading(true);
        let items = res.data;
        setData(data => [...items.forecast.forecastday])
        setLoading(false);
        setIsError(false);
        return res
      })
      .catch(error => {
        setLoading(false)
        setIsError(true)
        throw new Error(error)
      })
  }, [apiURL, isError]);

  if (isLoading) {
    return (
      <Box sx={{ display: 'flex' }}>
        <CircularProgress />
      </Box>
    )
  }
  if (isError && !isLoading) {
    return <ErrorBlock>Неверное название города</ErrorBlock>
  }

  return (
    <div>
      <WeatherBlock loading={isLoading} nameHead={townName} data={data} />
      <TemperatureChart loading={isLoading} data={data} />
    </div>
  )

};

export default WeatherGetData;