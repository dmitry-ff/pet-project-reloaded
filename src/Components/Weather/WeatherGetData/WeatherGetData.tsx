import React from 'react';
import axios from 'axios';
import TemperatureChart from '../TemperatureChart/TemperatureChart';
import { CircularProgress, Box } from '@mui/material';
import { TResponse } from "../types/TResponse";
import { TData } from "../types/TData";
import WeatherBlock from '../WeatherBlock/WeatherBlock';

import { EmotionJSX } from '@emotion/react/types/jsx-namespace';
import Chart from '../CanvasChart/CanvasChart';
import { Switch } from '@mui/material';
import * as Styled from './WeatherGetData.styled';
import { TWeatherGetData } from '../types/TWeatherGetData';



function WeatherGetData({ apiURL, townName }: TWeatherGetData): EmotionJSX.Element {

  const [data, setData] = React.useState<TData[]>([]);
  const [isLoading, setLoading] = React.useState<boolean>(true);
  const [isError, setIsError] = React.useState<boolean>(false);
  const [isSwitch, setIsSwitch] = React.useState<boolean>(true);

  React.useEffect(() => {
    axios.get<TResponse>(apiURL)
      .then(res => {
        setLoading(true);
        let items = res.data;
        setData(data => [...items.forecast.forecastday])
        setLoading(false);
        setIsError(false);
        return res
      })
      .catch(error => {
        setLoading(false)
        setIsError(true)
        throw new Error(error)
      })
  }, [apiURL, isError]);

  if (isLoading) {
    return (
      <Box sx={{ display: 'flex' }}>
        <CircularProgress />
      </Box>
    )
  }
  if (isError && !isLoading) {
    return <Styled.ErrorBlock>Неверное название города</Styled.ErrorBlock>
  }
  return (
    <div>
      <WeatherBlock loading={isLoading} nameHead={townName} data={data} />
      <Styled.SwitchDisplay>
        <span>Vanilla canvas</span>
        <Switch defaultChecked onClick={() => { setIsSwitch(isSwitch => !isSwitch) }} />
        <span>ChartJS</span>
      </Styled.SwitchDisplay>
      {isSwitch && <TemperatureChart loading={isLoading} data={data} />}
      {!isSwitch && <Chart loading={isLoading} data={data} />}

    </div >
  )

};

export default WeatherGetData;