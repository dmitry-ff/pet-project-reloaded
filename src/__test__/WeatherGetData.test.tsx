import { shallow } from "enzyme";
import renderer from 'react-test-renderer';
import WeatherBlock from '../Components/Weather/WeatherBlock'
import WeatherGetData from "../Components/Weather/WeatherGetData";
const mockProps = {
  apiURL: 'http://api.weatherapi.com/v1/forecast.json?key=3e2b23202a804627a5c115738222701&q=Tula&days=3&aqi=no&alerts=no',
  townName: 'Tula'
}
type TProps = {
  apiURL: string,
  townName: string
}
const setUp = (props: TProps) => shallow(<WeatherGetData {...props} />)
describe('WeatherGetData component', () => {
  let component;
  beforeEach(() => {
    component = setUp(mockProps);

  })
  it('WeatherGetData snapshot', () => {
    const tree = renderer
      .create(<WeatherGetData {...mockProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  })
  it('should render Weather component', () => {
    component = setUp(mockProps);
    const wrapper = component.find('Weather');
    expect(wrapper.length).toBe(1);
  })
  it('should render TemperatureChart component', () => {
    component = setUp(mockProps);
    const wrapper = component.find('TemperatureChart');
    expect(wrapper.length).toBe(1);
  })
})
export { }