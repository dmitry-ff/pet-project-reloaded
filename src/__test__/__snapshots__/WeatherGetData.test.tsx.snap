// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherGetData component WeatherGetData snapshot 1`] = `null`;
